src/Akka.Persistence.Redis/Query/Stages/AllEventsSource.cs

```diff
@@ -0,0 +1,343 @@
+                        _maxOffset = len;
+                        switch (_state)
+                        {
+                            case State.QueryWhenInitializing:
+                                // during initialization, downstream asked for an element,
+                                // let’s query elements
+                                _state = State.Idle;
+                                Query();
+                                break;
+                            case State.Initializing:
+                                // no request from downstream, just go idle
+                                _state = State.Idle;
+                                break;
+                            default:
+                                Log.Error($"Unexpected source state when initializing: {_state}");
+                                FailStage(new IllegalStateException($"Unexpected source state when initializing: {_state}"));
+                                break;
+                        }
+                    });
+
+                    _redis.GetDatabase(_database).ListLengthAsync(_journalHelper.GetEventsKey()).ContinueWith(task =>
+                    {
+                        if (!task.IsCanceled || task.IsFaulted)
+                        {
+                            initCallback(task.Result - 1);
+                        }
+                        else
+                        {
+                            Log.Error(task.Exception, "Error while initializing current events by tag");
+                            FailStage(task.Exception);
+                        }
+                    });
+                }
+            }
+
+            public override void PostStop()
+            {
+                _subscription?.UnsubscribeAll();
+            }
+
+            private void Query()
+            {
+                switch (_state)
+                {
+                    case State.Idle:
+                        if (_buffer.Count == 0)
+                        {
+                            // so, we need to fill this buffer
+                            _state = State.Querying;
+
+                            // request next batch of events for this tag (potentially limiting to the max offset in the case of non live stream)
+                            var refs = _redis.GetDatabase(_database).ListRange(_journalHelper.GetEventsKey(), _currentOffset, Math.Min(_maxOffset, _currentOffset + _max - 1));
+
+                            var trans = _redis.GetDatabase(_database).CreateTransaction();
+
+                            var events = refs.Select(bytes =>
+                            {
+                                var (sequenceNr, persistenceId) = bytes.Deserialize();
+                                return trans.SortedSetRangeByScoreAsync(_journalHelper.GetJournalKey(persistenceId), sequenceNr, sequenceNr);
+                            }).ToList();
+
+                            trans.ExecuteAsync().ContinueWith(task =>
```

Comment:
You are calling an async method, but I don't see any await on it. Is this intentional? At what point is the Task being executed?

Reply:
Yes, because `private void Query()` is not async and ContinueWith helps us to handle the `result` and any exception that might be thrown without awaiting

Reply:
I'm not really sure I understand what the code here does either - just looking at it

Reply:
Surprise surprise, I stole that code from:
https://github.com/akkadotnet/Akka.Persistence.Redis/blob/946aea895ff6ac2d8ee464948eae72ed3d4191a8/src/Akka.Persistence.Redis/Query/Stages/EventsByTagSource.cs#L290-L343

Reply:
What is the code doing?

For a live query, there is the need to prevent concurrent query to the server and since an actor processes messages one-by-one - sequentially, we can rely on `_state` to tell us when to call the server.

So if `_state is Idle && _buffer is empty` we need to set the `_state` to `Querying` since we are going to be calling the server in the background where we use the `_callback` object to `_buffer.Enqueue` but if the `_buffer` is not empty, we deliver/dequeue the messages first.

That is how I understand that code.

Reply:
I have no problem with the approach, but are you sure that the code was executed? Can you provide proof that it did?
A simple screenshot with a Console.WriteLine() would suffice.

Reply:
![image](https://user-images.githubusercontent.com/8049896/102904142-96fba700-4471-11eb-8dfc-b88507154151.png)